sdk/webpubsub/azure-messaging-webpubsub-client/src/main/java/com/azure/messaging/webpubsub/client/WebPubSubClientBuilder.java

```diff
@@ -0,0 +1,245 @@
+     * @return itself.
+     */
+    public WebPubSubClientBuilder retryOptions(RetryOptions retryOptions) {
+        this.retryOptions = retryOptions;
+        return this;
+    }
+
+    /**
+     * Allows for setting common properties such as application ID, headers, proxy configuration, etc.
+     *
+     * @param clientOptions A configured instance of {@link HttpClientOptions}.
+     * @return Returns the same concrete type with the appropriate properties updated, to allow for fluent chaining of
+     *      operations.
+     */
+    public WebPubSubClientBuilder clientOptions(ClientOptions clientOptions) {
+        this.clientOptions = clientOptions;
+        return this;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public WebPubSubClientBuilder configuration(Configuration configuration) {
+        this.configuration = configuration;
+        return this;
+    }
+
+    /**
+     * Sets whether automatically reconnect after disconnect.
+     *
+     * @param autoReconnect whether automatically reconnect after disconnect.
+     * @return itself.
+     */
+    public WebPubSubClientBuilder autoReconnect(boolean autoReconnect) {
+        this.autoReconnect = autoReconnect;
+        return this;
+    }
+
+    /**
+     * Sets whether automatically restore joined groups after reconnect.
+     *
+     * @param autoRestoreGroup whether automatically restore joined groups after reconnect.
+     * @return itself.
+     */
+    public WebPubSubClientBuilder autoRestoreGroup(boolean autoRestoreGroup) {
+        this.autoRestoreGroup = autoRestoreGroup;
+        return this;
+    }
+
+    /**
+     * Builds the client.
+     *
+     * @return the client.
+     */
+    public WebPubSubClient buildClient() {
+        return new WebPubSubClient(this.buildAsyncClient());
+    }
+
+    /**
+     * Builds the asynchronous client.
+     *
+     * @return the asynchronous client.
+     */
+    WebPubSubAsyncClient buildAsyncClient() {
+        // retryOptions
+        RetryStrategy retryStrategy;
+        if (retryOptions != null) {
+            if (retryOptions.getExponentialBackoffOptions() != null) {
+                retryStrategy = new ExponentialBackoff(retryOptions.getExponentialBackoffOptions());
+            } else if (retryOptions.getFixedDelayOptions() != null) {
+                retryStrategy = new FixedDelay(retryOptions.getFixedDelayOptions());
+            } else {
+                throw LOGGER.logExceptionAsError(
+                    new IllegalArgumentException("'retryOptions' didn't define any retry strategy options"));
+            }
+        } else {
+            retryStrategy = new ExponentialBackoff();
+        }
+
+        // credential
+        Mono<String> clientAccessUrlProvider;
+        if (credential != null && clientAccessUrl != null) {
+            throw LOGGER.logExceptionAsError(
+                new IllegalArgumentException("Both credential and clientAccessUrl have been set. "
```

Comment:
nit: we should use `IllegalStateException` here as the state of the builder (2 properties are non-null when one of them is expected to be null) is invalid. 
https://azure.github.io/azure-sdk/java_introduction.html#exceptions